pennylane/devices/default_clifford.py

```diff
@@ -519,58 +685,306 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+            for idx, wire in enumerate(meas_wires):
+                expectation = prob_sim.peek_z(wire)
+                # (Eig --> Res) | -1 --> 1 | 1 --> 0 | 0 --> 0 / 1 |
+                outcome = int(0.5 * (1 - expectation))
+                if not expectation:
+                    prob_res[tgt_index] /= 2.0
+                else:
+                    nope_idx = (
+                        np.where(
+                            np.squeeze(np.all(tgt_states[:, :idx] == tgt_state[:idx], axis=-1))
+                            & tgt_states[:, idx]
+                            != outcome
+                        )[0]
+                        if idx
+                        else np.where(tgt_states[:, idx] != outcome)[0]
+                    )
+                    nope_idx = np.setdiff1d(nope_idx, visited_probs)
+                    prob_res[nope_idx] = 0.0
+                    visited_probs.extend(tgt_integs[nope_idx])
+
+                    if tgt_state[idx] != outcome:
+                        prob_res[tgt_index] = 0.0
+                        break
+
+                prob_sim.postselect_z(wire, desired_value=tgt_state[idx])
+            visited_probs.append(tgt_integ)
+        return prob_res
+
+    def _sample_expectation(self, meas, stim_circuit, shots, seed):
+        """Measure the expectation value with respect to samples from simulator device."""
+        # Get the observable for the expectation value measurement
+        meas_op = meas.obs
+        if isinstance(meas_op, BasisStateProjector):
+            stim_circ = stim_circuit.copy()
+            stim_circ.append_from_stim_program_text("M " + " ".join(map(str, meas_op.wires)))
+            sampler = stim_circ.compile_sampler(seed=seed)
+            samples = qml.math.array(sampler.sample(shots=shots), dtype=int)
+            matches = np.where((samples == meas_op.data[0]).all(axis=1))[0]
+            return len(matches) / shots
+
+        samples, coeffs = self._measure_observable_sample(meas_op, stim_circuit, shots, seed)
+        expecs = [
+            qml.math.mean(qml.math.power([-1] * shots, qml.math.sum(sample, axis=1)))
+            for sample in samples
+        ]
+
+        return qml.math.dot(coeffs, expecs)
+
+    def _sample_variance(self, meas, stim_circuit, shots, seed):
+        """Measure the variance with respect to samples from simulator device."""
+        # Get the observable for the expectation value measurement
+        meas_op = meas.obs
+        meas_obs = qml.operation.convert_to_opmath(meas_op)
+        meas_obs1 = meas_obs.simplify()
+        meas_obs2 = (meas_obs1**2).simplify()
+
+        # use the naive formula for variance, i.e., Var(Q) = ⟨𝑄^2⟩−⟨𝑄⟩^2
+        return (
+            self._sample_expectation(qml.expval(meas_obs2), stim_circuit, shots, seed)
+            - self._sample_expectation(qml.expval(meas_obs1), stim_circuit, shots, seed) ** 2
+        )
+
+    @staticmethod
+    def _measure_single_sample(stim_ct, meas_ops, meas_idx, meas_wire):
```

Comment:
Does it merit a dedicated function? Could this be also `_measure_observable_sample`?

Reply:
Yes, this is specialized for efficiently calculating shadows. I would like to keep it separate.